src/execution/index/art/art_key.cpp

```diff
@@ -20,6 +21,186 @@ static uint8_t FlipSign(uint8_t key_byte) {
 	return key_byte ^ 128;
 }
 
+uint32_t EncodeFloat(float x)
+{
+    int shift;
+    unsigned long sign, exp, hibits, buff;
+    double fnorm, significand;
+    int expbits = 8;
+    int significandbits = 23;
+
+    //! zero (can't handle signed zero)
+    if (x == 0)
+    {
+        buff = 0;
+        return buff;
+    }
+    //! infinity
+    if (x > FLT_MAX)
+    {
+        buff = 128 + ((1 << (expbits - 1)) - 1);
+        buff <<= (31 - expbits);
+        return buff;
+    }
+    //! -infinity
+    if (x < -FLT_MAX)
+    {
+        buff = 128 + ((1 << (expbits - 1)) - 1);
+        buff <<= (31 - expbits);
+        buff |= (1 << 31);
+        return buff;
+    }
+    //! NaN
+    if (x != x)
+    {
+        buff = 128 + ((1 << (expbits - 1)) - 1);
+        buff <<= (31 - expbits);
+        buff |= 1234;
+        return buff;
+    }
+
+    //! get the sign
+    if (x < 0) { sign = 1; fnorm = -x; }
+    else { sign = 0; fnorm = x; }
+
+    //! get the normalized form of f and track the exponent
+    shift = 0;
+    while (fnorm >= 2.0) { fnorm /= 2.0; shift++; }
+    while (fnorm < 1.0) { fnorm *= 2.0; shift--; }
+
+    //! check for denormalized numbers
+    if (shift < -126)
+    {
+        while (shift < -126) { fnorm /= 2.0; shift++; }
+        shift = -1023;
+    }
+        //! out of range. Set to infinity
+    else if (shift > 128)
+    {
+        buff = 128 + ((1 << (expbits - 1)) - 1);
+        buff <<= (31 - expbits);
+        buff |= (sign << 31);
+        return buff;
+    }
+    else
+        fnorm = fnorm - 1.0; //! take the significant bit off mantissa
+
+    //! calculate the integer form of the significand
+    //! hold it in a  double for now
+
+    significand = fnorm * ((1LL << significandbits) + 0.5f);
+
+
+    //! get the biased exponent
+    exp = shift + ((1 << (expbits - 1)) - 1); //! shift + bias
+
+    hibits = (long)(significand);
+    buff = (sign << 31) | (exp << (31 - expbits)) | hibits;
+
+    return buff;
+}
+
+
+uint64_t EncodeDouble(double x) {
+    int shift;
+    unsigned long sign, exp, hibits, hilong, lowlong;
+    double fnorm, significand;
+    int expbits = 11;
+    int significandbits = 52;
+    uint64_t buff;
+    //! zero (can't handle signed zero)
+    if (x == 0)
+    {
+        buff = 0;
+        return buff;
+    }
+    //! infinity
+    if (x > DBL_MAX)
+    {
+        hilong = 1024 + ((1 << (expbits - 1)) - 1);
+        hilong <<= (31 - expbits);
+        lowlong = 0;
+        buff = hilong;
+        buff <<=32;
+        buff += lowlong;
+        return buff;
+    }
+    //! -infinity
+    if (x < -DBL_MAX)
+    {
+        hilong = 1024 + ((1 << (expbits - 1)) - 1);
```

Comment:
Could you add tests for -infinity, infinity and NaN? Perhaps you can generate them through e.g. multiplying large numbers, or dividing by zero.